src/pip/_internal/cli/base_command.py

```diff
@@ -87,6 +87,13 @@ def __init__(self, name, summary, isolated=False):
             self.parser,
         )
         self.parser.add_option_group(gen_opts)
+        try:
+            # mypy raises error due to
+            # https://github.com/python/mypy/issues/5868
+            self.add_options(self.cmd_opts)  # type: ignore
+        except AttributeError:
+            # it means that the base class has not defined the method
+            logger.debug("No add_options method defined in the class")
```

Comment:
Don't do this.

Instead, it would be better to define a `def add_options(self, cmd_opts): pass` on this class, so that inheriting classes that don't need to add additional options don't have to define this method.

Reply:
Okay! great then. I actually did all the changes as you suggested earlier. I also modified most of the test cases. The only place, where I was stuck was this here - https://github.com/pypa/pip/blob/master/tests/unit/test_base_command.py#L163, where `Command` class is directly being invoked. It raised an error with no `add_option` defined.

Reply:
It's more like the case with Abstract classes, where we actually don't create the object of such classes.